pyani_plus/workflows/__init__.py

```diff
@@ -20,3 +20,173 @@
+        )
+        raise ValueError(msg)
+
+    return input_files
+
+
+def progress_bar_via_db_comparisons(
+    database: Path, run_id: int, interval: float = 1.0
+) -> None:
+    """Show a progress bar based on monitoring the DB entries.
+
+    Will only self-terminate once all the run's comparisons are
+    in the database! Up to the caller to terminate it early.
+    """
+    session = db_orm.connect_to_db(database)
+    run = session.query(db_orm.Run).where(db_orm.Run.run_id == run_id).one()
+
+    already_done = run.comparisons().count()
+    total = run.genomes.count() ** 2 - already_done
+
+    done = 0
+    old_db_version = -1
+    with Progress(*progress_columns) as progress:
+        task = progress.add_task("Comparing pairs", total=total)
+        while done < total:
+            sleep(interval)
+            # Have there been any DB changes?
+            db_version = (
+                session.connection().execute(text("PRAGMA data_version;")).one()[0]
+            )
+            if old_db_version == db_version:
+                continue
+            old_db_version = db_version
+            new = run.comparisons().count() - already_done - done
+            progress.update(task, advance=new)
+            done += new
+    session.close()
+
+
+def run_snakemake_with_progress_bar(  # noqa: PLR0913
+    executor: ToolExecutor,
+    workflow_name: str,
+    targets: list[Path] | list[str],
+    params: dict,
+    working_directory: Path,
+    *,
+    show_progress_bar: bool = False,
+    database: Path | None = None,
+    run_id: int | None = None,
+    interval: float = 0.5,
+) -> None:
+    """Run snakemake with a progress bar.
+
+    Currently this works by monitoring the filesystem for the appearance of
+    the output files, while snakemake runs in a subprocess. This is not ideal,
+    and ideally we would use some kind of callbacks from snakemake - perhaps
+    using their logging mechanism.
+
+    The datatabase and run_id are only required with a progress bar.
+    """
+    import importlib.resources as impresources
+
+    from snakemake.cli import args_to_api, parse_args
+
+    from pyani_plus import workflows
+
+    if show_progress_bar and (database is None or run_id is None):
+        msg = "Both database and run_id are required with show_progress_bar=True"
+        raise ValueError(msg)
+
+    # Check this now to give clear up-front error - this function will be
+    # call from within the actual snakemake workflow files themselves:
+    if "indir" in params:
+        check_input_stems(params["indir"])
+
+    # Path to anim snakemake file
+    snakefile = impresources.files(workflows) / workflow_name
+
+    # Want snakemake to pull everything else from its profiles,
+    # most easily controlled via setting $SNAKEMAKE_PROFILE
+    parser, args = parse_args(
+        [
+            "--quiet",
+            "all" if show_progress_bar else "rules",
```

Comment:
Quiet by default is OK, but this needs to be controllable. For example adding a ``--verbose`` vs ``--quiet`` switch?